src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -231,7 +231,7 @@ def cache_metadata_if_selectable(
             # we should cache it on the child plan too.
             # This is necessary SelectStatement.select() will need the column states of the child plan
             # (check the implementation of derive_column_states_from_subquery().
-            if source_plan.projection is None:
+            if not source_plan.has_projection:
```

Comment:
yes, I have the same question not source_plan.has_projection meaning that if source_plan.has_projection is None or for example a empty list [], the statement would be true, is that desired behavior?

Reply:
`has_projection` is a new property I added for `SelectStatement` which is a check to know if we have added any projection on top of the child Selectable. Before this change, there was only one way to add projection which was with select method that updates `projection` attribute. Now, both existence of `projection` and `exclude_cols` mean we have added projection on top of child selectable.

https://github.com/snowflakedb/snowpark-python/pull/3316/files#diff-ba9463440d935e1d5667a2909f7d2b8061f56de7bdabd8207a29903e9eae8f59R859-R864

I can update this to ensure that `has_projection` is True when
- `projection is None`
- `projection == []`
- `exclude_cols is None`